How many late days are you using for this assignment? - 0

In [20]:
# Imports and pip installations (if needed)
import pandas as pd

pd.set_option('display.max_columns', None)

# Part 1: Load the dataset

In [21]:
# Load the given datasets
categorical_data = pd.read_csv("data/chronic_kidney_disease_categorical.csv")
numerical_data = pd.read_csv("data/chronic_kidney_disease_numerical.csv")

# Print the data
display(categorical_data)
display(numerical_data)

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,203694,38.0,80.0,99.0,19.0,0.5,147.0,3.5,13.6,44.0,7300.0,6.4
1,938027,43.0,60.0,108.0,25.0,1.0,144.0,5.0,17.8,43.0,7200.0,5.5
2,421471,37.0,60.0,111.0,35.0,0.8,135.0,4.1,16.2,50.0,5500.0,5.7
3,764115,70.0,90.0,144.0,125.0,4.0,136.0,4.6,12.0,37.0,8200.0,4.5
4,240975,47.0,80.0,95.0,35.0,0.9,140.0,4.1,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
415,522912,56.0,80.0,165.0,55.0,1.8,NaN,NaN,13.5,40.0,11800.0,5.0
416,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4
417,168242,48.0,60.0,112.0,44.0,1.2,142.0,4.9,14.5,44.0,9400.0,6.4
418,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

This is a dataset that provides features that can be used to predict chronic kidney disease. It includes data collected from a hospital over 2 months. 

The label in this dataset is the Target column, which represents whether the patient in that specific row had chronic kidney disease (ckd) or not (notckd). It is the answer/output that we wan to predict based on the other data, the features. The features in this dataset are all the columns except for the Target column (the label column), which has the data points that we use to predict the outcome. Features like age, blood pressure and red blood cells can be analyzed to see their corelation with the Target column, to see how well they can be used to predict whether someone has chronic kidney disease.

The features in the numerical dataset (numerical_data) are all numerical, while the features in the categorical dataset (categorical_data) are all categorical.

3 examples of categorical columns are:\
    1. Specific gravity\
    2. Sugar\
    3. Albumin

3 examples of numerical columns are:\
    1. Age\
    2. Blood pressure\
    3. Blood urea

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [22]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_dupes_size = len(numerical_data[numerical_data.duplicated()])
print(f'Number of duplicate rows in numerical_data: {num_dupes_size}')

# Drop these duplicate rows
numerical_data_deduped = numerical_data.drop_duplicates()

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_dupes_size = len(categorical_data[categorical_data.duplicated()])
print(f'Number of duplicate rows in categorical_data: {cat_dupes_size}')
categorical_data_deduped = categorical_data.drop_duplicates()


Number of duplicate rows in numerical_data: 30
Number of duplicate rows in categorical_data: 30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

No, all the unique IDs are not present in both the datasets. 

If we look at the number of rows in both datasets in part 1, the categorical dataset has 430 rows while the numerical dataset has 420 rows. Since each row has a unique id, this means that there are at least 10 unique ids in the categorical dataset that are not present in the numerical dataset. Additionally, it might be possible for a patient to only have categorical data or numerical data recorded and not the other, which could cause there to be unique ids in one table that aren't in the other.

In the combined dataset, if a certain row is missing data from the numerical dataset, for example, all columns that come from the numerical dataset would have NaN instead of a value (as shown in the first displayed dataframe below). It would be the same for the categorical dataset too.

In [23]:
# Checking how a row that is missing from one of datasets looks like 
merged_with_info = pd.merge(numerical_data_deduped, categorical_data_deduped, how = "outer", on="unique_id", indicator=True)
display(merged_with_info[merged_with_info['_merge'] == 'right_only'])

# Merge the two given numerical and categorical datasets based on their unique_ID.
merged = pd.merge(numerical_data_deduped, categorical_data_deduped, how = "outer", on="unique_id")

#Print the combined dataset
display(merged)

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target,_merge
47,208034,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,yes,yes,good,yes,no,ckd,right_only
58,240430,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,normal,normal,NaN,NaN,no,no,no,good,no,no,notckd,right_only
64,254775,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.0,2.0,normal,normal,notpresent,notpresent,yes,yes,yes,good,no,yes,ckd,right_only
69,266335,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,right_only
80,292312,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,right_only
83,298297,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,right_only
97,325238,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,poor,no,yes,ckd\t,right_only
173,488947,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3.0,0.0,abnormal,abnormal,notpresent,notpresent,no,no,no,good,no,no,ckd,right_only
261,663387,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,poor,no,yes,ckd,right_only
376,948317,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd,right_only


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,103570,24.0,80.0,125.0,NaN,NaN,136.0,3.5,15.4,43.0,5600.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,104675,65.0,80.0,215.0,133.0,2.5,NaN,NaN,13.2,41.0,NaN,NaN,2.0,1.0,normal,normal,present,notpresent,no,yes,no,good,no,no,ckd
3,105923,90.0,90.0,139.0,89.0,3.0,140.0,4.1,12.0,37.0,7900.0,3.9,1.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
4,107297,60.0,70.0,NaN,NaN,NaN,NaN,NaN,16.4,43.0,10800.0,5.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,11500.0,3.3,4.0,3.0,normal,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd
396,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,996412,70.0,60.0,120.0,40.0,0.5,140.0,4.6,16.0,43.0,4500.0,4.9,NaN,NaN,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
398,997629,35.0,NaN,99.0,30.0,0.5,135.0,4.9,15.4,48.0,5000.0,5.2,0.0,0.0,normal,normal,NaN,NaN,no,no,no,good,no,no,notckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [24]:
# Calculate the percentage of rows that contain atleast one missing value
percentage_with_atleast_one_missing = merged.isnull().any(axis=1).mean() * 100
# Print %
print(f'Percentage of rows that contain at least one missing value: {percentage_with_atleast_one_missing}')
# Drop these rows from the dataset
merged_without_nan = merged.dropna()
# Print the Dataset
print(merged_without_nan.columns)
merged_without_nan

Percentage of rows that contain at least one missing value: 61.75000000000001
Index(['unique_id', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo',
       'pcv', 'wbcc', 'rbcc', 'al', 'su', 'rbc', 'pc', 'pcc', 'ba', 'htn',
       'dm', 'cad', 'appet', 'pe', 'ane', 'Target'],
      dtype='object')


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
6,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
7,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
8,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,7100.0,5.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
9,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,6000.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
390,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,7900.0,3.9,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
391,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,10200.0,5.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
394,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,9800.0,6.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,11500.0,3.3,4.0,3.0,normal,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd


## Part 3.4: Sort the dataset according to the Labels

In [25]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_data = merged_without_nan.sort_values(by='Target').reset_index(drop=True)
# Print the dataset
sorted_data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,6700.0,3.9,4.0,0.0,normal,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,8800.0,3.4,3.0,4.0,normal,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [26]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
sorted_with_dummy_columns = pd.get_dummies(sorted_data, columns=categorical_columns, dtype=int)
# Print the dataset
sorted_with_dummy_columns

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,6700.0,3.9,4.0,0.0,0,1,1,0,0,1,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,8800.0,3.4,3.0,4.0,0,1,0,1,1,0,1,0,0,1,0,1,1,0,0,1,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 

This would be equivalent in downstream modelling if the single column had only 2 values (binary). This is because splitting it into 2 columns versus having only two options for values is both 2 dimensions, meaning it would have the same effect.

However, when adding more than 2 values, like the condiment example, this changes. Keeping those 5 different values within one column might cause later modelling to interpret that column as a numerical column rather than a categorical one since there would be more that 2 different values to analyze. It could also cause later analysis to  misinterpret the numbers that the condiments represent as rankings or order, which could affect the rating prediction. It could mislead the analysis into creating relations between these categories (condiments) due to similar numbers.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [27]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
means = sorted_with_dummy_columns[numerical_columns].mean()
std_devs = sorted_with_dummy_columns[numerical_columns].std()

mask = (sorted_with_dummy_columns[numerical_columns] <= means + (3 * std_devs)) & \
(sorted_with_dummy_columns[numerical_columns] >= means - (3 * std_devs))

mask = mask.all(axis=1)
sorted_without_outliers = sorted_with_dummy_columns.loc[mask, :].reset_index(drop=True)

sorted_without_outliers

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,397388,63.0,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
2,438182,73.0,80.0,253.0,142.0,4.6,138.0,5.8,10.5,33.0,7200.0,4.3,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
3,474407,21.0,90.0,107.0,40.0,1.7,125.0,3.5,8.3,23.0,12400.0,3.9,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
4,475200,64.0,100.0,163.0,54.0,7.2,140.0,4.6,7.9,26.0,7500.0,3.4,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
132,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [28]:
# Normalize the all Numerical Attributes in the dataset.
normalized = sorted_without_outliers.copy()
normalized[numerical_columns] = (normalized[numerical_columns] - normalized[numerical_columns].min()) / (normalized[numerical_columns].max() - normalized[numerical_columns].min())
# Print the dataset
normalized

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,343710,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,0.153846,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,397388,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.171429,0.606557,0.631579,0.443548,0.410256,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
2,438182,0.905405,0.50,0.785408,0.862745,0.518519,0.600000,0.828571,0.401639,0.447368,0.233871,0.435897,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
3,474407,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,0.333333,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
4,475200,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.485714,0.188525,0.263158,0.258065,0.205128,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,486397,0.189189,0.25,0.227468,0.222222,0.074074,0.500000,0.257143,0.737705,0.736842,0.096774,0.564103,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,203466,0.216216,0.00,0.115880,0.052288,0.098765,0.600000,0.400000,0.647541,0.684211,0.290323,0.974359,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,482105,0.513514,0.00,0.107296,0.235294,0.012346,0.600000,0.371429,0.770492,0.894737,0.274194,0.948718,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
132,228849,0.702703,0.50,0.128755,0.261438,0.098765,0.666667,0.171429,0.688525,0.894737,0.193548,1.000000,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

The unique_id column is not appropriate for modelling and predictions. This column is only used as a "primary key" or "index" to uniquely identify each row. It does not provide any useful information towards making predictions. It could even mislead the analysis into trying to find connections between this unique_id and Target prediction, when there aren't any.

In [29]:
#Remove that column
without_id = normalized.drop(columns=['unique_id'])
without_id = without_id.reset_index(drop=True)
# Print the dataset
without_id

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,0.153846,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.171429,0.606557,0.631579,0.443548,0.410256,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
2,0.905405,0.50,0.785408,0.862745,0.518519,0.600000,0.828571,0.401639,0.447368,0.233871,0.435897,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
3,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,0.333333,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
4,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.485714,0.188525,0.263158,0.258065,0.205128,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,0.189189,0.25,0.227468,0.222222,0.074074,0.500000,0.257143,0.737705,0.736842,0.096774,0.564103,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,0.216216,0.00,0.115880,0.052288,0.098765,0.600000,0.400000,0.647541,0.684211,0.290323,0.974359,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,0.513514,0.00,0.107296,0.235294,0.012346,0.600000,0.371429,0.770492,0.894737,0.274194,0.948718,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
132,0.702703,0.50,0.128755,0.261438,0.098765,0.666667,0.171429,0.688525,0.894737,0.193548,1.000000,0.0,0.0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [30]:
# Export the dataframe to a new csv file
without_id.to_csv("data/cleaned_data.csv")
# Export the dataframe to a new json file
without_id.to_json("data/cleaned_data.json")

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [31]:
without_id.head(15)

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.743243,0.50,0.442060,0.901961,0.432099,0.500000,0.657143,0.172131,0.210526,0.395161,0.153846,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.171429,0.606557,0.631579,0.443548,0.410256,2.0,2.0,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
2,0.905405,0.50,0.785408,0.862745,0.518519,0.600000,0.828571,0.401639,0.447368,0.233871,0.435897,2.0,0.0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
3,0.202703,0.75,0.158798,0.196078,0.160494,0.166667,0.171429,0.221311,0.184211,0.653226,0.333333,4.0,0.0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
4,0.783784,1.00,0.399142,0.287582,0.839506,0.666667,0.485714,0.188525,0.263158,0.258065,0.205128,4.0,2.0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
5,0.729730,0.00,0.935622,0.169935,0.160494,0.333333,0.028571,0.188525,0.236842,0.879032,0.102564,3.0,1.0,0,1,1,0,0,1,1,0,0,1,1,0,1,0,0,1,1,0,0,1,1,0
6,0.675676,0.75,0.253219,0.633987,0.777778,0.366667,0.542857,0.286885,0.342105,0.169355,0.205128,2.0,0.0,1,0,1,0,1,0,1,0,0,1,1,0,1,0,1,0,1,0,1,0,1,0
7,0.851351,0.25,0.618026,0.562092,0.728395,0.000000,0.285714,0.311475,0.315789,0.580645,0.179487,4.0,3.0,0,1,1,0,0,1,0,1,0,1,0,1,0,1,1,0,0,1,0,1,1,0
8,0.540541,0.00,0.399142,0.535948,0.358025,0.700000,0.314286,0.344262,0.315789,0.830645,0.153846,1.0,0.0,0,1,0,1,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0
9,0.756757,0.25,0.223176,0.209150,0.160494,0.533333,0.514286,0.573770,0.605263,0.290323,0.333333,3.0,0.0,0,1,1,0,1,0,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0


Prompt:

Convert this dataframe to a markdown table. Make sure to do this for every row from 1 to 14. Leave the label for the first column blank.

age bp bgr bu sc sod pot hemo pcv wbcc rbcc al su rbc_abnormal rbc_normal pc_abnormal pc_normal pcc_notpresent pcc_present ba_notpresent ba_present htn_no htn_yes dm_no dm_yes cad_no cad_yes appet_good appet_poor pe_no pe_yes ane_no ane_yes Target_ckd Target_notckd 0 0.743243 0.50 0.442060 0.901961 0.432099 0.500000 0.657143 0.172131 0.210526 0.395161 0.153846 2.0 0.0 1 0 1 0 1 0 1 0 0 1 0 1 0 1 0 1 0 1 0 1 1 0 1 0.770270 1.00 0.901288 0.163399 0.345679 0.766667 0.171429 0.606557 0.631579 0.443548 0.410256 2.0 2.0 0 1 0 1 1 0 0 1 0 1 1 0 0 1 1 0 1 0 1 0 1 0 2 0.905405 0.50 0.785408 0.862745 0.518519 0.600000 0.828571 0.401639 0.447368 0.233871 0.435897 2.0 0.0 1 0 1 0 1 0 1 0 0 1 0 1 0 1 1 0 1 0 1 0 1 0 3 0.202703 0.75 0.158798 0.196078 0.160494 0.166667 0.171429 0.221311 0.184211 0.653226 0.333333 4.0 0.0 0 1 1 0 0 1 0 1 1 0 1 0 1 0 1 0 1 0 0 1 1 0 4 0.783784 1.00 0.399142 0.287582 0.839506 0.666667 0.485714 0.188525 0.263158 0.258065 0.205128 4.0 2.0 1 0 1 0 1 0 0 1 0 1 0 1 1 0 1 0 0 1 1 0 1 0 5 0.729730 0.00 0.935622 0.169935 0.160494 0.333333 0.028571 0.188525 0.236842 0.879032 0.102564 3.0 1.0 0 1 1 0 0 1 1 0 0 1 1 0 1 0 0 1 1 0 0 1 1 0 6 0.675676 0.75 0.253219 0.633987 0.777778 0.366667 0.542857 0.286885 0.342105 0.169355 0.205128 2.0 0.0 1 0 1 0 1 0 1 0 0 1 1 0 1 0 1 0 1 0 1 0 1 0 7 0.851351 0.25 0.618026 0.562092 0.728395 0.000000 0.285714 0.311475 0.315789 0.580645 0.179487 4.0 3.0 0 1 1 0 0 1 0 1 0 1 0 1 0 1 1 0 0 1 0 1 1 0 8 0.540541 0.00 0.399142 0.535948 0.358025 0.700000 0.314286 0.344262 0.315789 0.830645 0.153846 1.0 0.0 0 1 0 1 1 0 1 0 0 1 0 1 1 0 1 0 1 0 1 0 1 0 9 0.756757 0.25 0.223176 0.209150 0.160494 0.533333 0.514286 0.573770 0.605263 0.290323 0.333333 3.0 0.0 0 1 1 0 1 0 1 0 0 1 0 1 1 0 1 0 1 0 1 0 1 0 10 0.662162 0.50 0.618026 0.411765 0.432099 0.566667 0.571429 0.434426 0.473684 0.250000 0.282051 3.0 1.0 0 1 1 0 0 1 0 1 0 1 0 1 1 0 1 0 0 1 1 0 1 0 11 0.783784 0.00 0.725322 0.313725 0.481481 0.566667 0.714286 0.319672 0.342105 0.258065 0.205128 4.0 1.0 1 0 1 0 1 0 0 1 0 1 0 1 1 0 0 1 0 1 1 0 1 0 12 0.878378 0.00 0.206009 0.751634 0.604938 0.533333 0.571429 0.475410 0.500000 0.879032 0.435897 4.0 0.0 0 1 0 1 1 0 1 0 0 1 0 1 1 0 0 1 0 1 1 0 1 0 13 0.878378 0.25 0.639485 0.470588 0.395062 0.433333 0.428571 0.393443 0.447368 0.104839 0.256410 3.0 0.0 0 1 1 0 0 1 0 1 0 1 0 1 0 1 1 0 1 0 1 0 1 0 14 0.716216 0.50 1.000000 0.163399 0.111111 0.066667 0.171429 0.393443 0.500000 0.532258 0.435897 1.0 0.0 1 0 0 1 1 0 1 0 1 0 0 1 1 0 0 1 1 0 1 0 1 0

### Paste the markdown table here

|   | age     | bp   | bgr       | bu        | sc        | sod      | pot      | hemo     | pcv      | wbcc     | rbcc     | al   | su   | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|---|---------|------|-----------|-----------|-----------|----------|----------|----------|----------|----------|----------|------|------|---------------|-------------|--------------|------------|----------------|--------------|----------------|------------|--------|---------|-------|--------|--------|---------|-------------|------------|-------|--------|--------|---------|-------------|----------------|
| 0 | 0.743243 | 0.50 | 0.442060  | 0.901961  | 0.432099  | 0.500000 | 0.657143 | 0.172131 | 0.210526 | 0.395161 | 0.153846 | 2.0  | 0.0  | 1             | 0           | 1            | 0          | 1              | 0            | 1              | 0          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1          | 0     | 1      | 0      | 1       | 0           | 1              |
| 1 | 0.770270 | 1.00 | 0.901288  | 0.163399  | 0.345679  | 0.766667 | 0.171429 | 0.606557 | 0.631579 | 0.443548 | 0.410256 | 2.0  | 2.0  | 0             | 1           | 0            | 1          | 1              | 0            | 0              | 1          | 0      | 1       | 1     | 0      | 0      | 1       | 1           | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 2 | 0.905405 | 0.50 | 0.785408  | 0.862745  | 0.518519  | 0.600000 | 0.828571 | 0.401639 | 0.447368 | 0.233871 | 0.435897 | 2.0  | 0.0  | 1             | 0           | 1            | 0          | 1              | 0            | 1              | 0          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1          | 1     | 0      | 1      | 0       | 1           | 0              |
| 3 | 0.202703 | 0.75 | 0.158798  | 0.196078  | 0.160494  | 0.166667 | 0.171429 | 0.221311 | 0.184211 | 0.653226 | 0.333333 | 4.0  | 0.0  | 0             | 1           | 1            | 0          | 0              | 1            | 0              | 1          | 1      | 0       | 1     | 0      | 1      | 0       | 1           | 0          | 1     | 0      | 0      | 1       | 1           | 0              |
| 4 | 0.783784 | 1.00 | 0.399142  | 0.287582  | 0.839506  | 0.666667 | 0.485714 | 0.188525 | 0.263158 | 0.258065 | 0.205128 | 4.0  | 2.0  | 1             | 0           | 1            | 0          | 1              | 0            | 0              | 1          | 0      | 1       | 0     | 1      | 1      | 0       | 1           | 0          | 0     | 1      | 1      | 0       | 1           | 0              |
| 5 | 0.729730 | 0.00 | 0.935622  | 0.169935  | 0.160494  | 0.333333 | 0.028571 | 0.188525 | 0.236842 | 0.879032 | 0.102564 | 3.0  | 1.0  | 0             | 1           | 1            | 0          | 0              | 1            | 1              | 0          | 0      | 1       | 1     | 0      | 0      | 1       | 1           | 0          | 0     | 1      | 1      | 0       | 0           | 1              |
| 6 | 0.675676 | 0.75 | 0.253219  | 0.633987  | 0.777778  | 0.366667 | 0.542857 | 0.286885 | 0.342105 | 0.169355 | 0.205128 | 2.0  | 0.0  | 1             | 0           | 1            | 0          | 1              | 0            | 1              | 0          | 0      | 1       | 1     | 0      | 1      | 0       | 1           | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 7 | 0.851351 | 0.25 | 0.618026  | 0.562092  | 0.728395  | 0.000000 | 0.285714 | 0.311475 | 0.315789 | 0.580645 | 0.179487 | 4.0  | 3.0  | 0             | 1           | 1            | 0          | 0              | 1            | 0              | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1          | 1     | 0      | 1      | 0       | 1           | 0              |
| 8 | 0.540541 | 0.00 | 0.399142  | 0.535948  | 0.358025  | 0.700000 | 0.314286 | 0.344262 | 0.315789 | 0.830645 | 0.153846 | 1.0  | 0.0  | 0             | 1           | 0            | 1          | 1              | 0            | 1              | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 1           | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 9 | 0.756757 | 0.25 | 0.223176  | 0.209150  | 0.160494  | 0.533333 | 0.514286 | 0.573770 | 0.605263 | 0.290323 | 0.333333 | 3.0  | 0.0  | 0             | 1           | 1            | 0          | 1              | 0            | 1              | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 1           | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 10| 0.662162 | 0.50 | 0.618026  | 0.411765  | 0.432099  | 0.566667 | 0.571429 | 0.434426 | 0.473684 | 0.250000 | 0.282051 | 3.0  | 1.0  | 0             | 1           | 1            | 0          | 0              | 1            | 0              | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 1           | 0          | 1     | 0      | 0      | 1       | 1           | 0              |
| 11| 0.783784 | 0.00 | 0.725322  | 0.313725  | 0.481481  | 0.566667 | 0.714286 | 0.319672 | 0.342105 | 0.258065 | 0.205128 | 4.0  | 1.0  | 1             | 0           | 1            | 0          | 1              | 0            | 0              | 1          | 0      | 1       | 0     | 1      | 1      | 0       | 1           | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 12| 0.878378 | 0.00 | 0.206009  | 0.751634  | 0.604938  | 0.533333 | 0.571429 | 0.475410 | 0.500000 | 0.879032 | 0.435897 | 4.0  | 0.0  | 0             | 1           | 0            | 1          | 1              | 0            | 1              | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 1           | 0          | 1     | 0      | 0      | 1       | 1           | 0              |
| 13| 0.878378 | 0.25 | 0.639485  | 0.470588  | 0.395062  | 0.433333 | 0.428571 | 0.393443 | 0.447368 | 0.104839 | 0.256410 | 3.0  | 0.0  | 0             | 1           | 1            | 0          | 0              | 1            | 0              | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1          | 1     | 0      | 1      | 0       | 1           | 0              |
| 14| 0.716216 | 0.50 | 1.000000  | 0.163399  | 0.111111  | 0.066667 | 0.171429 | 0.393443 | 0.500000 | 0.532258 | 0.435897 | 1.0  | 0.0  | 1             | 0           | 0            | 1          | 1              | 0            | 1              | 0          | 1      | 0       | 1     | 0      | 0      | 0       | 1           | 0          | 1     | 0      | 1      | 0       | 1           | 0              |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer:

I faced these hallucinations while trying to generate the markdown table. The generation would commonly stop halfway and not generate all 15 rows of the table. It took multiple tries to finally get the 383GPT to generate the whole markdown table. I have not checked all the cells to ensure that the data is accurate, but I would not be surprised if some of them were wrong since LLMs tend to get confused with a lot of data.

I think the main issue is that, when provided with a lot of data, LLMs tend to get confused about what to focus on. There are a few ways I have found to mitigate this:
1. Enclose the data being given in some kind of delimiters like ``` and """ or use other formatting techniques to highlight which sections are the data. This lets the model differentiate between the instructions and the data and focus on all parts of the instructions without getting lost in the data.

2. Highlighting instructions by adding indications for what is an instruction can help the LLM focus on what it needs to do.

3. You could add post processing steps to get the LLM to evaluate its own response to see whether it matches the instructions given in the prompt.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [32]:
# Code to rename all the columns in the dataset

mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the DataFrame columns
renamed_data = sorted_data.rename(columns=mapping, inplace=False)
renamed_data

,unique_id,age,blood pressure,blood glucose random,blood urea,serum creatinine,sodium,potassium,hemoglobin,packed cell volume,wbcc,rbcc,albumin,sugar,red blood cells,pus cell,pus cell clumps,bacteria,hypertension,diabetes mellitus,coronary artery disease,appetite,pedal edema,anemia,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,6700.0,3.9,4.0,0.0,normal,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,8800.0,3.4,3.0,4.0,normal,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [33]:
# Converted SQL to Pandas code

result = without_id.groupby('Target_ckd').size().reset_index(name='count')
result

,Target_ckd,count
0,0,111
1,1,23


*Describe what the above code does here*

Answer:

This code looks at all the values in the Target_ckd column of the dataframe 'without_id'. It groups the dataframe by the 'Target_ckd' column and returns the dataframe with contains the frequencies of all of the values in 'Target_ckd'. The returned dataframe's 'Target_ckd' column indicates the value for which the frequency is given in the count column.